### PR TITLE
Assignment2 - Johanne McClenahan and Alexa Darrington

### PR DESCRIPTION
**Pull request by JohanneJayde**
@alyxadarr - Alexa's GitHub ID
@JohanneJayde  - Johanne's GitHub ID

### CODE REVIEW

**Comment by SergeiAshikhmin** [comment]
Reviewing :)

File: Logger/FileLogger.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        public override void Log(LogLevel logLevel, string message)
+        {
+
+            using StreamWriter streamWrite = File.AppendText(PathName);
+
+            string log = DateTime.Now.ToString(CultureInfo.InvariantCulture) + " " + ClassName + " "
+                + logLevel.ToString() + ": " + message;
+
+            streamWrite.WriteLine(log);
+
```
Comment (SergeiAshikhmin) [inline Logger/FileLogger.cs:26]: If I remember correctly, we have to close the StreamWriter every time.
 ```suggestion
streamWrite.Close();
```

File: Logger/LogFactory.cs [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
+
+    public BaseLogger? CreateLogger(string className)
     {
+        if (_pathName == null)
+        {
+            return null;
+        }
 
-        return null;
+        return new FileLogger(_pathName) {
```
Comment (SergeiAshikhmin) [inline Logger/LogFactory.cs:18]: Braces should be on a new line.
```suggestion
        return new FileLogger(_pathName) 
        {
```

**Review by SergeiAshikhmin — COMMENTED** [review]
Great job, the code looks clean and intuitive. 👍

There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ 
Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔

File: Logger/FileLogger.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        public FileLogger(string pathName)
+        {
+            PathName = pathName;
+        }
+
+        public override void Log(LogLevel logLevel, string message)
+        {
+
+            using StreamWriter streamWrite = File.AppendText(PathName);
```
Comment (rkdeboer) [inline Logger/FileLogger.cs:20]: I also used AppendText! Nice work!

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 4]
```diff
@@ -10,30 +10,171 @@ public class BaseLoggerMixinsTests
...
+        BaseLoggerMixins.Warning(logger, message);
+
+        // Assert
+        Assert.AreEqual(1, logger.LoggedMessages.Count);
+        Assert.AreEqual(LogLevel.Warning, logger.LoggedMessages[0].LogLevel);
+        Assert.AreEqual(expectedMessage, logger.LoggedMessages[0].Message);
+    }
+
+    [TestMethod]
+    [DataRow("Intellij is installed on computer number 468 at Alpha Station", "{0} is installed on computer number {1} at {2} Station")]
```
Comment (rkdeboer) [inline Logger.Tests/BaseLoggerMixinsTests.cs:79]: I like that you tested with more complex messages!

**Comment by rkdeboer** [comment]
Very well done! 
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 4 above)
Comment (JohanneJayde) [reply Logger.Tests/BaseLoggerMixinsTests.cs:79]: Thanks!

File: Logger/FileLogger.cs (hunk 3 above)
Comment (JohanneJayde) [reply Logger/FileLogger.cs:20]: Nice, we tried using WriteAllText before realizing we are supposed to append to the end of the file

File: Logger.Tests/FileLoggerTests.cs [hunk 5]
```diff
@@ -20,20 +20,33 @@ public void Log_ValidMessage_WritesToFileCorrectly(string expectedContent, LogLe
     {
 
         // Arrange
-        string filePath = string.Join(@"\", (Environment.CurrentDirectory));
-        string path = Path.Combine(filePath, fileName);
-        LogFactory logFactory = new();
-        logFactory.ConfigureFileLogger(path);
-        FileLogger? fileLogger = logFactory.CreateLogger(nameof(FileLoggerTests)) as FileLogger;
+        MockFileLogger fileLogger = new(fileName);
```
Comment (HadiM2) [inline Logger.Tests/FileLoggerTests.cs:23]: Great Addition on the mock logger!

File: Logger.Tests/FileLoggerTests.cs [hunk 6]
```diff
@@ -20,20 +20,33 @@ public void Log_ValidMessage_WritesToFileCorrectly(string expectedContent, LogLe
...
-        logFactory.ConfigureFileLogger(path);
-        FileLogger? fileLogger = logFactory.CreateLogger(nameof(FileLoggerTests)) as FileLogger;
+        MockFileLogger fileLogger = new(fileName); 
         string loggedMessageWithDate = $"{DateTime.Now.ToString(CultureInfo.InvariantCulture)} {expectedContent}";
 
         // Act
-        fileLogger?.Log(level, message);
-        string fileContents = File.ReadLines(path).Last();
+        fileLogger.Log(level, message);
+        string fileContents = File.ReadLines(GetCorrectPath(fileName)).Last();
```
Comment (HadiM2) [inline Logger.Tests/FileLoggerTests.cs:28]: Great job on making the code cleaner!

**Review by HadiM2 — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
Great Unit Testing, Good Job!

File: Logger.Tests/FileLoggerTests.cs (hunk 6 above)
Comment (JohanneJayde) [reply Logger.Tests/FileLoggerTests.cs]: Thanks Hadi!

File: Logger.Tests/FileLoggerTests.cs (hunk 5 above)
Comment (JohanneJayde) [reply Logger.Tests/FileLoggerTests.cs]: Thanks! I tried to follow what they suggested in class :)

**Pull request closed without merging**